### DISCUSSION TITLE
"<module_name>" is not accessed Pylance warning when using Conda

### DISCUSSION THREAD

**Discussion by Ritesh313** [microsoft/pylance-release#5143] [Q&A] [answered]
I'm getting `"<module_name>"is not accessed` warnings from Pylance in my python file while importing all the stdlib modules.
The warning description just says: 
`conda_env_location/lib/python3.9/<module_name>.py" is overriding the stdlib module "<module_name>"`

I have my anaconda environment selected as my interpreter in VS code. I know it's just a warning and not interrupting my code, but I would just like to know if I have something setup wrong. Please let me know if I should provide some other information for this.

Modules which have this warning: datetime, random, typing, argparse
Platform: VS code
Environment: Conda
Python version: 3.9.7

Here's a screenshot of the warning:
<img width="1475" alt="image" src="https://github.com/microsoft/pylance-release/assets/36135489/f59ab3c3-fe40-46ae-96ae-d7da0fcb6905">

**Comment by EricYoung37** [comment]
**EDIT: Ignore this question. It's just a mental lapse. Sorry for that.**

Having similar issue while I'm using a native Python interpreter (that is directly installed on my PC).

In my case, I have a module `transaction_manager.py` that contains `class TransactionManager`. However, even if I've imported the module and instantiate an object of the class in the module, Pylance complains that the module is not accessed and the class is not defined.

![image](https://github.com/microsoft/pylance-release/assets/77303738/16668886-ea6d-4cd1-bc2b-f5e6884e600f)

**Reply by rchiodo** [reply]
@EricYoung7 I believe the correct way to reference TransactionManager would be like so:

```python
tm = transaction_manager.TransactionManager()
```

Otherwise, you're not using the module.

**Reply by EricYoung37** [reply]
Thanks for your response! Sorry for the dumb question. I haven't been using Python for a while and I guess I was thinking stupid last night as I was working late.

I was wondering the same thing. It seemed to make sense that I got access to everything in a module by just importing it without specifying it in front of the member of the module. However, I googled and then realized specification is indeeded necessary if `from import` is not being used. I find it reasonable since referencing two methods having the same name from different modules may be ambiguous without specificication of the module.

**Reply by rchiodo** [reply]
No worries, I had to look it up myself to be sure. 

I believe this is why it behaves that way: ` I find it reasonable since referencing two methods having the same name from different modules may be ambiguous without specificication of the module.`

**Comment by rchiodo** [comment]
@Ritesh313 it looks like your configuration thinks the site-packages folder is user code for some reason. We'd need to see your settings.json and the log as described here:
https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#filing-an-issue

**Reply by Ritesh313** [reply]
@rchiodo I have since updated my conda enviroment. Here's a requirements.txt for the new environment: 
[requirements.txt](https://github.com/microsoft/pylance-release/files/13531183/requirements.txt)


here's my settings.json:
```
{   "python.analysis.autoImportCompletions": true,
    "python.analysis.logLevel": "Trace",
    "python.analysis.fixAll": ["source.unusedImports"],  
    "editor.defaultFormatter": "ms-python.black-formatter",
    "files.exclude": {
        "**/__pycache__": true,
        "**/.cache": true,
        "**/.coverage": true,
        "**/.coverage.*": true,
        "**/.hypothesis": true,
        "**/.mypy_cache": true,
        "**/.nox": true,
        "**/.pytest_cache": true,
        "**/.ruff_cache": true,
        "**/.tox": true},
    "editor.indentSize": "tabSize",
    "editor.detectIndentation": false,
    "files.autoSave": "afterDelay",
    "github.copilot.enable": {
        "*": true,
        "plaintext": false,
        "markdown": false,
        "scminput": false}}
```
Python Language Server logs a significant amount and gets truncated, so I have copied the first part of the log here. Let me know if you'll need the complete log, I can increase the limit and attach it in a file here. 
```
2023-12-01 14:00:43.786 [info] [Info  - 2:00:43 PM] (40003) Pylance language server 2023.11.10 (pyright 088ebaa5) starting
2023-12-01 14:00:43.786 [info] [Info  - 2:00:43 PM] (40003) Server root directory: /home/riteshchowdhry/.vscode-server-insiders/extensions/ms-python.vscode-pylance-2023.11.10/dist
2023-12-01 14:00:43.795 [info] [Info  - 2:00:43 PM] (40003) Starting service instance "riteshchowdhry"
2023-12-01 14:00:43.850 [info] (40003) No configuration file found.
2023-12-01 14:00:43.850 [info] (40003) No pyproject.toml file found.
2023-12-01 14:00:43.850 [info] [Info  - 2:00:43 PM] (40003) Setting pythonPath for service "riteshchowdhry": "/blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/bin/python"
2023-12-01 14:00:43.850 [info] [Info  - 2:00:43 PM] (40003) Setting environmentName for service "riteshchowdhry": "3.11.6 (pt2 conda)"
2023-12-01 14:00:44.026 [info] [Info  - 2:00:44 PM] (40003) Assuming Python version 3.11
2023-12-01 14:00:44.027 [info] (40003) Assuming Python platform Linux
2023-12-01 14:00:45.241 [info] [Info  - 2:00:45 PM] (40003) Search paths for /blue/azare/riteshchowdhry
2023-12-01 14:00:45.243 [info] [Info  - 2:00:45 PM] (40003)   /home/riteshchowdhry/.vscode-server-insiders/extensions/ms-python.vscode-pylance-2023.11.10/dist/typeshed-fallback/stdlib
2023-12-01 14:00:45.248 [info] [Info  - 2:00:45 PM] (40003)   /blue/azare/riteshchowdhry
2023-12-01 14:00:45.291 [info] [Info  - 2:00:45 PM] (40003)   /blue/azare/riteshchowdhry/typings
2023-12-01 14:00:45.363 [info] [Info  - 2:00:45 PM] (40003)   /home/riteshchowdhry/.vscode-server-insiders/extensions/ms-python.vscode-pylance-2023.11.10/dist/typeshed-fallback/stubs/...
2023-12-01 14:00:45.395 [info] [Info  - 2:00:45 PM] (40003)   /home/riteshchowdhry/.vscode-server-insiders/extensions/ms-python.vscode-pylance-2023.11.10/dist/bundled/stubs
2023-12-01 14:00:45.396 [info] [Info  - 2:00:45 PM] (40003)   /blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11
2023-12-01 14:00:45.397 [info] [Info  - 2:00:45 PM] (40003)   /blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11/lib-dynload
2023-12-01 14:00:45.397 [info] [Info  - 2:00:45 PM] (40003)   /blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11/site-packages
2023-12-01 14:00:45.397 [info] [Info  - 2:00:45 PM] (40003) Adding fs watcher for library directories:
 /blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11
/blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11/lib-dynload
/blue/azare/riteshchowdhry/virt_envs_anaconda/pt2/lib/python3.11/site-packages
2023-12-01 14:00:45.397 [info] [Info  - 2:00:45 PM] (40003) Adding fs watcher for directories:
 /blue/azare/riteshchowdhry
2023-12-01 14:00:45.398 [info] (40003) Searching for source files
2023-12-01 14:00:47.490 [info] [Info  - 2:00:47 PM] (40003) Auto-excluding /blue/azare/riteshchowdhry/Personal/scale_ai/chowdhry_ritesh_take_home
2023-12-01 14:00:54.664 [info] [Error - 2:00:54 PM] (40003) Enumeration of workspace source files is taking longer than 10 seconds.
This may be because:
* You have opened your home directory or entire hard drive as a workspace
* Your workspace contains a very large number of directories and files
* Your workspace contains a symlink to a directory with many files
* Your workspace is remote, and file enumeration is slow
To reduce this time, open a workspace directory with fewer files or add a pyrightconfig.json configuration file with an "exclude" section to exclude subdirectories from your workspace. For more details, refer to https://github.com/microsoft/pyright/blob/main/docs/configuration.md.

```

**Reply by rchiodo** [reply]
Are you still getting the same error? Your env path in conda looks to be this:

`/blue/azare/riteshchowdhry/virt_envs_anaconda/pt2`

Does the error show this path or the original one?

**Reply by rchiodo** [reply]
That's also a strange folder path. It's like you made a venv by hand in the `/blue/azare/riteshchowdhry/virt_envs_anaconda/pt2` folder? Did you make that by hand or did anaconda do that? (Maybe they've changed how they generate envs from the last time I used anaconda)

**Reply by Ritesh313** [reply]
I'm getting the same warning even for the new environment 
![Screenshot (7)](https://github.com/microsoft/pylance-release/assets/36135489/e015106b-a73a-4a2a-8ce5-e6d7c70e1304)
About the path: I'm using a SLURM server and by default anaconda installs virtual environment in `home/user/` directory, but we have limited space there and it's not recommended to store big files there. Instead I have it installed in a storage drive on that server `/blue/azare/riteshchowdhry/`

**Reply by rchiodo** [reply]
How are you accessing the drive? Is there a symlink to it by any chance? I think this sounds like a bug on our part, transferring this to an issue.

**Reply by Ritesh313** [reply]
I have a symlink to it in the home directory, but I'm not using that. There are 2 ways for me to access it: `/blue/azare/riteshchowdhry/virt_envs_anaconda` or `/home/riteshchowdhry/blue/virt_envs_anaconda`. I'm using the former as my workspace in VS code. I'm using a remote tunnel to VS code and using that drive as my workspace in vs code. I can try using it through the symlink in my home directory and see if that changes anything.

**Reply by rchiodo** [reply]
What's in your root folder for `virt_envs_anaconda`. I'm wondering if this maybe messing up our virtual environment detection. That would also cause this problem.

**Reply by rchiodo** [reply]
Can you reproduce this is you don't make your workspace root the directory above where the virtual environment is? Meaning so that the `pt2` folder isn't listed in the workspace?

**Reply by rchiodo** [reply]
Yeah, this may just be a bug in our virtual environment detection. That's why this error is coming out:

```
2023-12-01 14:00:54.664 [info] [Error - 2:00:54 PM] (40003) Enumeration of workspace source files is taking longer than 10 seconds.
This may be because:
* You have opened your home directory or entire hard drive as a workspace
* Your workspace contains a very large number of directories and files
* Your workspace contains a symlink to a directory with many files
* Your workspace is remote, and file enumeration is slow
To reduce this time, open a workspace directory with fewer files or add a pyrightconfig.json configuration file with an "exclude" section to exclude subdirectories from your workspace. For more details, refer to https://github.com/microsoft/pyright/blob/main/docs/configuration.md.
```

**Reply by rchiodo** [reply]
That would happen if we thought all of the files in the virtual workspace were user files.

**Reply by rchiodo** [reply]
It's this code here that checks if a folder should be skipped or not:
https://github.com/microsoft/pyright/blob/23942f31e902a82084960611e0db2051a86f999b/packages/pyright-internal/src/analyzer/service.ts#L1121C15-L1121C26

It might be a conda environment doesn't have a bin/activate folder.

**Reply by rchiodo** [reply]
Yeah that's it. I can repro if I open my conda env folder as the root of the workspace:

![image](https://github.com/microsoft/pylance-release/assets/19672699/f534a283-3aca-4934-98f9-7db3fa000c55)

This is probably fixable on our end (detect this situation) but I don't think it's likely to be fixed any time soon. It's not a very common thing for people to have their conda environments exist in their workspace.

**Reply by Ritesh313** [reply]
There are indeed a lot of directories in the root folder:

![image](https://github.com/microsoft/pylance-release/assets/36135489/4787ebef-5184-4d3e-ba3d-7e4b8baec3dc)


I didn't understand what you mean by 
>  Meaning so that the pt2 folder isn't listed in the workspace?

This is where pt2 is (along with other conda environments I have)
![image](https://github.com/microsoft/pylance-release/assets/36135489/c2f42c72-8c94-4ac1-8564-936bde858d8d)

And I checked the pt2 folder. It doesn't have `bin/activate` folder.

Here's the .condarc file that I have in my home directory to make this (conda environment at a specific location) work:
```
channels:
  - conda-forge
  - bioconda
  - defaults
  - https://repo.anaconda.com/pkgs/main
  - https://repo.anaconda.com/pkgs/free
pkgs_dirs:
  - /blue/azare/riteshchowdhry/virt_envs_anaconda/pkgs/
auto_activate_base: false
anaconda_upload: false
env_prompt: ({name})
envs_dirs:
  - /blue/azare/riteshchowdhry/virt_envs_anaconda/
```

**Reply by rchiodo** [reply, marked as answer]
Yeah that's the root cause, your virtual env folder is in the workspace. So we treat it as user source because we can't tell that it's actually a virtual env directory. 

You can workaround this a number of ways:

1. Don't add your virtual env folder under your workspace. Put it somewhere else. You can still set it as the interpreter in use, but then Pylance won't parse it.
2. Add an explicit `python.analysis.exclude` setting for that directory
3. Add a `bin/activate` script to the venv directories. It doesn't actually have to do anything.

**Reply by Ritesh313** [reply]
Thanks! I added that directory to `python.analysis.exclude` in VS code settings. Discussion on excluding paths in VS code #2581

**Comment by rchiodo** [comment]
Made an issue out of this:
https://github.com/microsoft/pylance-release/issues/5191
